### PR TITLE
Ignore custom pretrained weights when resuming training

### LINKED ISSUE #25812: Resume loads the original custom `pretrained` weights instead of the resume checkpoint
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Issue title

`resume=True` silently overwrites resume checkpoint weights when `pretrained` is a path string

### Summary

When a training run is started with a custom pretrained checkpoint path, for example
`pretrained="path/to/pretrained.pt"`, that string is saved in the training checkpoint arguments. On resume,
`BaseTrainer.check_resume()` restores the saved `pretrained` string. `BaseTrainer.setup_model()` then loads the resume
checkpoint first, but immediately replaces its model weights with the original `pretrained` checkpoint.

The optimizer, scaler, EMA, epoch, and best fitness are still restored from the resume checkpoint. The resulting state
is therefore inconsistent:

| State | Loaded from |
| --- | --- |
| Training model weights | Original custom `pretrained.pt` |
| Optimizer state | Resume `last.pt` |
| AMP scaler state | Resume `last.pt` |
| EMA state | Resume `last.pt` |
| Epoch and best fitness | Resume `last.pt` |

This can cause an immediate loss jump and sustained accuracy degradation after resuming.

### Affected condition

The issue occurs when all of the following are true:

1. The original run used `pretrained` as a `str` or `Path`, not the usual Boolean `True`.
2. That value was saved in `last.pt` as part of `train_args`.
3. Training is resumed with `resume=True` or `resume=path/to/last.pt`.

The normal `pretrained=True` case does not enter the problematic `isinstance(..., (str, Path))` branch.

### Minimal reproduction

Start a small run from a YAML architecture while explicitly supplying a pretrained path:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.yaml")
model.train(
    data="coco8.yaml",
    epochs=10,
    pretrained="yolo11n.pt",
)
```

Stop the run after at least one completed epoch so that `last.pt` contains a resumable optimizer state. Then resume it:

```python
from ultralytics import YOLO

model = YOLO("runs/detect/train/weights/last.pt")
model.train(resume=True)
```

The resumed trainer reads `pretrained="yolo11n.pt"` from the saved checkpoint arguments. During
`BaseTrainer.setup_model()`, `load_checkpoint(last.pt)` is called first and `load_checkpoint(yolo11n.pt)` is called
afterward. The latter value becomes the `weights` argument passed to `get_model()`.

The problem can also be verified without relying on metrics by instrumenting or temporarily logging the two
`load_checkpoint()` calls in `BaseTrainer.setup_model()`.

### Relevant code path

Upstream references:

- [`BaseTrainer.setup_model()` on `main`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/trainer.py)
- [Commit `79b2086bfd` that moved custom `pretrained` loading outside the `.pt` branch](https://github.com/ultralytics/ultralytics/commit/79b2086bfd76e12f2b1bd2aa611ff437fbbcd457)

`BaseTrainer.check_resume()` restores the checkpoint arguments and only reinstates the resume model path:

```python
ckpt_args = load_checkpoint(last)[0].args
self.args = get_cfg(ckpt_args)
self.args.model = self.args.resume = str(last)
```

The saved string-valued `pretrained` argument remains in `self.args`. It is not one of the resume override keys.

`BaseTrainer.setup_model()` then performs both loads:

```python
if str(self.model).endswith(".pt"):
    weights, ckpt = load_checkpoint(self.model)  # resume last.pt
    cfg = weights.yaml
if isinstance(self.args.pretrained, (str, Path)):
    weights, _ = load_checkpoint(self.args.pretrained)  # overwrites resume weights
```

Finally, `resume_training(ckpt)` restores checkpoint state but does not restore the training model weights:

```python
self._load_checkpoint_state(ckpt)  # optimizer, scaler, EMA, best_fitness
```

Passing `pretrained=False` in the resume call is not currently a reliable workaround because `check_resume()` replaces
the current arguments with `ckpt_args`, and `pretrained` is not included in its allowed resume override keys.

### Expected behavior

When `resume` is active, the training model weights and all resumable training state should come from the same resume
checkpoint. A saved custom `pretrained` path should only describe how the original run was initialized; it should not be
loaded again during resume.

### Actual behavior

The training model is rebuilt from the original custom pretrained checkpoint while the optimizer and training progress
are restored from `last.pt`.

### Proposed fix

Do not load `args.pretrained` when resuming:

```diff
-        if isinstance(self.args.pretrained, (str, Path)):
+        if not self.resume and isinstance(self.args.pretrained, (str, Path)):
             weights, _ = load_checkpoint(self.args.pretrained)
```

A regression test could create a resumable checkpoint whose `train_args["pretrained"]` is a different checkpoint path,
then assert that `setup_model()` passes the resume checkpoint model—not the original pretrained model—to `get_model()`.

### Environment observed

- Ultralytics: `8.4.95`
- Repository commit: `d5a1e9472d37b658cada273a86f271c73e57dd54`
- Relevant behavior is also present in the current upstream `main` implementation of `BaseTrainer.setup_model()` at the
  time of reporting.
- OS, Python, and PyTorch versions: please fill in from the reproduction environment.

### Environment

OS: Linux-5.4.0-150-generic-x86_64-with-glibc2.35
Python: 3.11.11 | packaged by conda-forge | (main, Dec  5 2024, 14:17:24) [GCC 13.3.0]
Ultralytics: 8.4.95
PyTorch: 2.6.0+cu118
CUDA: 11.8
cuDNN: 90100
GPU: NVIDIA GeForce RTX 3090

### Minimal Reproducible Example

from ultralytics import YOLO

# Step 1: Start a new training run with `pretrained` specified as a path string.
model = YOLO("yolo11n.yaml")
model.train(
    data="coco8.yaml",
    epochs=10,
    pretrained="yolo11n.pt",
    project="runs/reproduce",
    name="resume_pretrained_bug",
)

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Ajaysingh-2003**
Fixes #25812

`check_resume()` restores every arg from the checkpoint, so a string-valued
`pretrained` path from the original run survives into the resumed run. In
`setup_model()` it then reloads over the weights just loaded from `last.pt`,
while the optimizer, scaler, EMA and epoch still come from the checkpoint —
training silently continues from reset weights.

The guard already exists on the sibling branch for `pretrained is False`;
this applies it to the string case too.

Also restores distillation resume, which lost its `DistillationModel`
whenever a pretrained path replaced the checkpoint weights.

## Verification

Trained with `pretrained=<path>`, interrupted after one epoch, stamped the
checkpoint weights with a distinct signature, then resumed:

| | before | after |
|---|---|---|
| loads on resume | `last.pt`, then `pretrained.pt` | `last.pt` only |
| resumed weights | pretrained | checkpoint |

`start_epoch=1` in both cases, confirming the inconsistent state.

Extended `test_setup_model_respects_pretrained_arg_for_pt_models` with a
resume case. The existing fake `load_checkpoint` returned the same object
for every path, so it had to become path-aware — otherwise a resume case
passes even with the bug present.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Ensure custom pretrained weights are ignored when resuming training from a checkpoint.

### 📊 Key Changes
- Updated `BaseTrainer.setup_model()` to skip loading `args.pretrained` when `self.resume` is enabled.
- Preserved existing behavior for normal training with string or `Path` pretrained weights and for explicitly disabled pretrained weights.
- Expanded engine tests to verify checkpoint, custom pretrained, and resume scenarios.

### 🎯 Purpose & Impact
- Prevents custom pretrained weights from overriding the checkpoint weights during resumed training.
- Makes resume behavior more predictable and ensures training continues from the saved checkpoint state.
- Adds regression coverage for the updated weight-selection logic.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ajaysingh-2003, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated welcome message, and an Ultralytics engineer will assist with your PR if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean, focused fix. The resume guard prevents checkpoint-restored string/Path `pretrained` values from replacing `last.pt` weights, while preserving the existing non-resume pretrained behavior. The parameterized regression test covers the affected resume case and existing true/false/path cases.

**Comment by glenn-jocher** [comment]
The fix and regression coverage look good; the only remaining blocker is the [CLA check](https://docs.ultralytics.com/help/CLA). Both `@Ajaysingh-2003` and `@Ajaysingh-2003-pgm` must comment exactly: `I have read the CLA Document and I sign the CLA`.

**Comment by Ajaysingh-2003** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Ajaysingh-2003-pgm** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
The current CLA check still lists @Ajaysingh-2003 and @Ajaysingh-2003-pgm as unsigned. @Ajaysingh-2003 @Ajaysingh-2003-pgm, please each post this exact comment on this PR so the CLA can be recorded:

> I have read the CLA Document and I sign the CLA

You have both posted this previously, but the latest CLA check still does not recognize the signatures, so please post it again.

**Comment by Ajaysingh-2003** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Ajaysingh-2003-pgm** [comment]
I have read the CLA Document and I sign the CLA
